File: 0001-0500/0261_validTree.py
class UnionFind:
    def __init__(self, n):
        self.parent = [node for node in range(n)]
        
    def find(self, node):
        while node != self.parent[node]:
            node = self.parent[node]
            
        return node
        
    def union(self, nodeA, nodeB):
        rootA = self.find(nodeA)
        rootB = self.find(nodeB)
        
        if rootA == rootB:
            return False
            
        self.parent[rootB] = rootA
        
        return True
        
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges) != n - 1:
            return False
            
        uf = UnionFind(n)
        
        for a, b in edges:
            if not uf.union(a, b):
                return False

# Alternative solution

class UnionFind:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.cc = n

    def find(self, a):
        while a != self.parent[a]:
            a = self.parent[a]
        return a

    def union(self, a, b):
        root_a = self.find(a)
        root_b = self.find(b)

        if root_a != root_b:
            self.parent[root_b] = root_a
            self.cc -= 1
            return True

class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        uf = UnionFind(n)

        for node1, node2 in edges:
            if not uf.union(node1, node2):
                return False        

        return uf.cc == 1
        
        return True

File: 0001-0500/0261_validTree.java
class UnionFind {
    int[] parent;
    
    public UnionFind(int n) {
        this.parent = new int[n];
        for (int i = 0; i < n; i++) {
            this.parent[i] = i;
        }
    }

    public int find(int a) {
        while (a != this.parent[a]) {
            a = this.parent[a];
        }
        return a;
    }

    public boolean union(int a, int b) {
        int aRoot = find(a);
        int bRoot = find(b);
        if (aRoot == bRoot) {
            return false;
        }
        this.parent[bRoot] = aRoot;
        return true;
    }
}

class Solution {
    public boolean validTree(int n, int[][] edges) {
        if (edges.length != n-1) {
            return false;
        }
        
        UnionFind uf = new UnionFind(n);
        for (int i = 0; i < edges.length; i++) {
            if (!uf.union(edges[i][0], edges[i][1])) {
                return false;
            }
        }
        
        return true;
    }
}
